gsc/_env.scm

```diff
@@ -153,24 +159,17 @@
 
 
 (define (env-lookup env name stop-at-first-frame? proc)
-
-  (define (assoc-set! al k v)
-    (let ((cell (assoc k al)))
-      (if cell
-          (begin
-            (set-cdr! cell v)
-            al)
-          (cons (cons k v) al))))
-
   (define (search env name full?)
     (if full?
         (search* env name #t)
         (let ((full-name-and-ns (env-namespace-lookup-both env name)))
           (if full-name-and-ns
-              (begin
-                (env-var-origins-set! env (assoc-set! (env-var-origins-ref env)
-                                                      (car full-name-and-ns)
-                                                      (cdr full-name-and-ns)))
+              (let ((full-name (car full-name-and-ns))
```

Comment:
For better error messages, it would seem better to have a table indexed by the namespace that contains for each namespace a set of full-names of the global referred to in that namespace (which could be represented with a table). That way all the undefined names can be grouped according to the namespace, which will make the error messages easier to understand.